qmetaobject/src/qtdeclarative.rs

```diff
@@ -502,16 +502,105 @@ pub fn qml_register_type<T: QObject + Default + Sized>(
     })
 }
 
-/// Wrapper around [`void qmlRegisterModule(const char *uri, int versionMajor, int versionMinor)`][qt] function.
+/// Register the given type as an uncreatable QML type
 ///
-/// [qt]: https://doc.qt.io/qt-5/qqmlengine.html#qmlRegisterModule
-#[cfg(qt_5_9)]
-pub fn qml_register_module(
+/// Refer to the Qt documentation for QQmlEngine::qmlRegisterUncreatableType.
+pub fn qml_register_uncreatable_type<T: QObject + Sized>(
     uri: &CStr,
     version_major: u32,
     version_minor: u32,
+    qml_name: &CStr,
+    no_creation_reason: QString,
 ) {
     let uri_ptr = uri.as_ptr();
+    let qml_name_ptr = qml_name.as_ptr();
+    let meta_object = T::static_meta_object();
+
+    extern "C" fn extra_destruct(c: *mut c_void) {
+        cpp!(unsafe [c as "QObject *"] {
+            QQmlPrivate::qdeclarativeelement_destructor(c);
+        })
+    }
+
+    let size = T::cpp_size();
+
+    let type_id = <RefCell<T> as PropertyType>::register_type(Default::default());
+
+    cpp!(unsafe [
+        qml_name_ptr as "char *",
+        uri_ptr as "char *",
+        version_major as "int",
+        version_minor as "int",
+        meta_object as "const QMetaObject *",
+        size as "size_t",
+        type_id as "int",
+        no_creation_reason as "QString"
+    ] {
+        // BEGIN: From QML_GETTYPENAMES
+        // FIXME: list type?
+        /*const int listLen = int(strlen("QQmlListProperty<"));
+        QVarLengthArray<char,64> listName(listLen + nameLen + 2);
+        memcpy(listName.data(), "QQmlListProperty<", size_t(listLen));
+        memcpy(listName.data()+listLen, className, size_t(nameLen));
+        listName[listLen+nameLen] = '>';
+        listName[listLen+nameLen+1] = '\0';*/
+        // END
+
+        int parserStatusCast = meta_object && qmeta_inherits(meta_object, &QQuickItem::staticMetaObject)
```

Comment:
I was just following qml_register_type. I replace this with -1, which seems to be the default. Can you confirm that it's correct?

Reply:
Ping, can you confirm that the default value of -1 is correct here?

Reply:
I don't know. The Qt code does put a value there but i don't know if it is usefull.
The type might not be created by QML but maybe it can still be cast?  I have no idea